### ISSUE TITLE
How we use SymbTR with MelodyRNN(Magenta)?

### ISSUE THREAD

**Issue by hedonistrh** [hedonistrh/TurkishMusicGeneration#26]
**Reason**
- [MelodyRNN](https://github.com/tensorflow/magenta/blob/master/magenta/models/melody_rnn/melody_rnn_model.py) currently does not work with SymbTR dataset. 
- MusicXML does not work.
- Half of the .midi files give _polyphony exception._

**Design**
- Try to understand which files work which do not.
- If we can not figure out any relation, we can try to manipulate source code for polyphony exception.
- At last step, we can use our LSTM implementation with same hyper-parameter of MagentaRNN.

**Acceptance Criteria**
- Other people should re-produce the result.

**Outcome**
- Our deliverable will be a method to work with MagentaRNN

**Comment by hedonistrh** [comment]
- _huzzam--sarki--aksak--vefa_yoktur--tanburi_mustafa_cavus.mid_  works **ok**
- _hicaz--sarki--curcuna--ruya_gibi--avni_anil.mid_ works **ok**
- _hicaz--sarki--aksak--gormedim_omrumun--kadri_sencalar.mid_ gives **exception**
- _huzzam--sarki--curcuna--aksam_donusu--udi_marko_colakoglu.mid_ works **ok**
- _karcigar--sarki--semai--vah_meyus-i--mahmud_celaleddin_pasa.mid_ gives **exception**

**Comment by sertansenturk** [comment]
Some of the scores in the above list are not part of the dataset. Can we
restrict this ticket to the scores so we use your efforts efficiently?

**Comment by hedonistrh** [comment]
I am trying to get understand the reason behind the problem. So, I think, at that step, it does not matter so much. For the next step, I will stick with our scores. 💯

**Comment by hedonistrh** [comment]
I will try to read midi files via [Music21](http://web.mit.edu/music21/) as I did before for [my _bestekar_ project.](https://github.com/hedonistrh/bestekar/blob/master/preprocess.ipynb) Maybe, we can understand the reason with that. 🖌

**Comment by hedonistrh** [comment]
Most interesting part is that why [_huzzam--sarki--curcuna--aksam_donusu--udi_marko_colakoglu.mid_](https://github.com/MTG/SymbTr/blob/master/midi/huzzam--sarki--curcuna--aksam_donusu--udi_marko_colakoglu.mid) works and gives no exception. 🤔 

![image](https://user-images.githubusercontent.com/34948815/53305797-1fb15800-3886-11e9-9264-6ffa7c05d8cd.png)

Also, when I check with _music21_, I realized that it contains two note at some time.

![image](https://user-images.githubusercontent.com/34948815/53305983-2345de80-3888-11e9-8e7e-4a6f2c2572a9.png)

**Comment by hedonistrh** [comment]
I just changed the source code of [Magenta's melody converter.](https://github.com/tensorflow/magenta/blob/master/magenta/music/melodies_lib.py)

line237 was:
```python
was ignore_polyphonic_notes=False 
```
new:
```python
ignore_polyphonic_notes=True
```

However, it does not change the result. I will focus this tomorrow. 🛌

**Comment by magdalenafuentes** [comment]
Hi @hedonistrh 

I'm a bit lost about what you could successfully do and what you couldn't:
Did you reproduce their results with any of magenta's models? (i.e. generate a melody with base_nn?) 
When you say the model fails, it means it's skipping files and training with a few files? Or not training at all?
Are you following [this?](https://github.com/tensorflow/magenta/blob/master/magenta/scripts/README.md) 
And finally, what about [what @sertansenturk said a about modifying the files?]( https://github.com/hedonistrh/TurkishMusicGeneration/pull/24#issuecomment-465365626)

**Comment by hedonistrh** [comment]
Hey @magdalenafuentes , thanks for comment. I agree with you. This is quite confusing. :)

- Exactly, I am following Magenta guide. :100: 
- We can generate melody with their dataset. 
- With SymbTr's midi, we can also generate. But, some of SymbTr files gives _polyphony_ exception. I am trying to understand reason behind that. If we can, hopefully, we can solve the problem. If it does not work, we will use our implementation.

In addition to that, I will just use our determined scores and check how many of them will give this exception.

**Comment by magdalenafuentes** [comment]
Thanks @hedonistrh !

It will be useful to know the numbers, i.e. the amount of files that launch the exception over total files (always talking about the files in @sertansenturk [dataset](https://github.com/sertansenturk/otmm_generation_dataset)). 

About converting MIDI/MusicXML to Note Sequence, are you doing that? Is there where you get the exceptions? Or what are you using as input to the model?

**Comment by hedonistrh** [comment]
- We can not use MusicXML, because of different accidental in Makam Music.
- We got that exception, when I was trying midi as input to **create dataset** command.

**Comment by hedonistrh** [comment]
Results from **selected midis**

![image](https://user-images.githubusercontent.com/34948815/53367646-03c1bb00-3947-11e9-931c-23d65ea1781b.png)


I will try to understand **source code**. If we can not figure out the solution until end of Wednesday, maybe we should skip Magenta RNN at that point. 🛌

**Comment by hedonistrh** [comment]
We should focus on [melody rnn's pipeline](https://github.com/tensorflow/magenta/blob/master/magenta/models/melody_rnn/melody_rnn_pipeline.py). But, the docker file contains old scripts which is from 2016 or 2017. 😪

I modified:

- [x] melody_pipelines.py
- [x] melody_rnn_create_dataset.py

But, it gives some exception. 😞 Maybe, we can manipulate our midi files via _music21_ 👍 I think, our best movement can be that :)

@sertansenturk @magdalenafuentes _**As a summary, I think we can try to modify our midis via music21. We can delete one of the note when two of them occur at the same time. What do you think?**_

**Comment by sertansenturk** [comment]
As before: please give it a try but do not lose a lot of time.

If music21 does not work, the page Magdalena has pointed mentions that we
can use the ABC notation too. As we talked before Seymour Shilen has
converted all symbtr scores to the ABC format.

If that does not work we have to write our own code convert the note
information in Symbtr-txt scores to Magenta's noteSequence. If it's going
to be the case, I think we have to discuss the feasibility given the time.

**Comment by magdalenafuentes** [comment]
I agree with @sertansenturk. I would pick literally a couple of examples that the code skips, modify them with music21, and see if it works. If it does, do the same with the rest. If not, let us know and we can think of something else. 

On top of that, a couple of questions:
1-  I don't completely understand the image you're sharing. Is the code being able to correctly parse 415 melodies and skipping 210? If so, isn't that enough to train the model and get some first results and see how we're doing?
2 - Each melody corresponds to a file here?

**Comment by barisbozkurt** [comment]
Just a suggestion: investing in music21 may result in time loss. Tweaking it is not straight-forward, it is very much western music oriented/designed. It is very nicely packed in an object oriented style, one needs to dismount many components to make modifications (unless you agree to build everything on top of western music). ABC would be a better option. If that does not work, writing new file read functions to meet input format would be better investment than spending time with music21.

**Comment by hedonistrh** [comment]
Thanks for the suggestion. 

@sertansenturk I would like to try music21 firstly. After that, we can talk about ABC notation.

@magdalenafuentes 
- Each melody corresponds one midi file
- I have not trained it. When I have time, I can try it.
- It parse 451 melodies(midi file) and produces 185 outputs. Other melodies gives different exceptions. (mostly polyphony)

@barisbozkurt What is the reason behind this small polyphonic parts in SymbTr dataset.

![image](https://user-images.githubusercontent.com/34948815/53305797-1fb15800-3886-11e9-9264-6ffa7c05d8cd.png)

**Comment by barisbozkurt** [comment]
Where do you observe ployphony? if in music21, it could be a defect. I don't remember any issue related to polyphony.  You can open that midi file in audacity and the check the points where you observe multi-pitch

**Comment by hedonistrh** [comment]
@barisbozkurt I am using Garageband to see midi files. One of the polyphonic segment in pink box.

![image](https://user-images.githubusercontent.com/34948815/53436039-83f52880-39fa-11e9-8d8c-ba996e0cd15b.png)

**Comment by sertansenturk** [comment]
I suspect this is pitch bends for tuning adjustments. Either Garageband
cannot cope with it or that is a mistake during midi conversion.

I suspect the first because I have never heard such jumps when I listen to
the symbtr scores.

Halil, can you check if the same jump happens at the same location in
audacity too? Also can you indicate the score & the time stamp the jump
happens?

**Comment by hedonistrh** [comment]
I think, same problem also occurs at Audacity.

<img width="862" alt="screen shot 2019-02-26 at 21 17 25" src="https://user-images.githubusercontent.com/34948815/53443430-fcb0b080-3a0b-11e9-80d9-36b5db2f436a.png">

**Comment by sertansenturk** [comment]
Can you write the score name & the timestamp?

I think it's not a pitch bends but it's an embellishment (çarpma - hammer
on/off). I'd like to verify it personally...

**Comment by hedonistrh** [comment]
[_huzzam--sarki--curcuna--aksam_donusu--udi_marko_colakoglu.mid_](https://github.com/MTG/SymbTr/blob/master/midi/huzzam--sarki--curcuna--aksam_donusu--udi_marko_colakoglu.mid) 

There are many timestamp in this file with the issue. 🦌

**Comment by umutsimsekli** [comment]
They don't seem to be very important, can you get rid of them easily?

**Comment by hedonistrh** [comment]
_karcigar--sarki--semai--vah_meyus-i--mahmud_celaleddin_pasa.mid_ gives exception, however, I can not realized that any polyphony in the file with Garageband and music21. 🤔 

Now, I think, we have 2 option

- [ ] Try use our LSTM model with same hyper-parameter of Magenta via our .txt files.

- [ ] Try to use ABC notation with MagentaRNN.

<img width="619" alt="screen shot 2019-02-26 at 23 10 26" src="https://user-images.githubusercontent.com/34948815/53450224-ee6a9080-3a1b-11e9-9af5-cb235eae618c.png">

**Comment by barisbozkurt** [comment]
Sertan is right, very likely it could be "çarpma" and they can be discarded.

**Comment by hedonistrh** [comment]
According to their [source code](https://github.com/tensorflow/magenta/blob/master/magenta/music/midi_io.py) in their repo, they are using [pretty-midi](https://github.com/craffel/pretty-midi). Also, I will check our datas with that library. :+1:

**Comment by hedonistrh** [comment]
List the files which gives exception or not. And their investigation via music21.

**Files gives exception**
- ussak--turku--sofyan--mevlam_bircok--malatya.mid _acc. to music21, polyphonic_
- rast--turku--sofyan--adananin_yollari--.mid _acc. to music21, polyphonic_
- hicaz--ilahi--duyek--icimde_bir--amir_ates.mid _acc. to music21, monophonic_
- huzzam--aranagme--duyek--1--.mid _acc. to music21, polyphonic_
- ussak--turku--aksak--kizilciklar_oldu--edirne-kesan.mid _acc. to music21, monophonic_
- segah--turku--curcuna--havuz_basinin--erzincan.mid _acc. to music21, monophonic_
- ussak--sarki--duyek--sevgi_dolu--bilge_ozgen.mid _acc. to music21, monophonic_
- nihavent--sarki--sofyan--kandilli_yuzerken--munir_nurettin_selcuk.mid _acc. to music21, monophonic_

**Files works ok**
- nihavent--sarki--duyek--sen_benim--emin_ongan.mid. _acc. to music21, monophonic_
- mahur--sarki--aksak--acirim_asik--giriftzen_asim_bey.mid _acc. to music21, monophonic_

- [ ] We have the list now. So, lets close #30.
- [ ] So, manipulation via music21 won't work. So, lets close #29 that issue.
- [ ] 

I will focus #31.

**Comment by hedonistrh** [comment]
_I wrote a script to check how pretty-midi visualize our datas._

**From my perspective, I can not understand why some of them works and some of them gives exception.**

**Results:**
_Gives Exception_
![image](https://user-images.githubusercontent.com/34948815/53668346-ead14680-3c73-11e9-96b0-f163f14103c9.png)

> - There are 1 time signature changes
> - There are 1 instruments
> - Instrument 0 has 400 notes
> - Instrument 0 has 400 pitch bends

![image](https://user-images.githubusercontent.com/34948815/53668364-02103400-3c74-11e9-92f7-25a6021d3f64.png)

> - There are 1 time signature changes
> - There are 1 instruments
> - Instrument 0 has 314 notes
> - Instrument 0 has 314 pitch bends

_Works **ok**_
![image](https://user-images.githubusercontent.com/34948815/53668543-6632f800-3c74-11e9-9f3b-25b840ce54bc.png)

> - There are 1 time signature changes
> - There are 1 instruments
> - Instrument 0 has 464 notes
> - Instrument 0 has 464 pitch bends

![image](https://user-images.githubusercontent.com/34948815/53668593-9aa6b400-3c74-11e9-9fad-d9354f4d2095.png)

> - There are 1 time signature changes
> - There are 1 instruments
> - Instrument 0 has 304 notes
> - Instrument 0 has 304 pitch bends

**Script:**
``` python
import wget
import pretty_midi
import os
import numpy as np
# For plotting
import mir_eval.display
import librosa.display
import matplotlib.pyplot as plt
import sys

def plot_piano_roll(pm, start_pitch, end_pitch, fs=100):
    # Use librosa's specshow function for displaying the piano roll
    librosa.display.specshow(pm.get_piano_roll(fs)[start_pitch:end_pitch],
                             hop_length=1, sr=fs, x_axis='time', y_axis='cqt_note',
                             fmin=pretty_midi.note_number_to_hz(start_pitch))

def midi_visualizer(filename):
	if (os.path.isfile(filename) == False):
		url = "https://github.com/MTG/SymbTr/raw/master/midi/"+filename
		wget.download(url) 

	pm = pretty_midi.PrettyMIDI(filename)
	print('There are {} time signature changes'.format(len(pm.time_signature_changes)))
        print('There are {} instruments'.format(len(pm.instruments)))
        print('Instrument 0 has {} notes'.format(len(pm.instruments[0].notes)))
        print('Instrument 0 has {} pitch bends'.format(len(pm.instruments[0].pitch_bends)))
	plt.figure(figsize=(12, 4))
	plot_piano_roll(pm, 48, 92)
	plt.title(filename)
	plt.show()
if __name__ == "__main__":
	midi_visualizer(sys.argv[1])
``` 

**Example usage:**
``` sh
python midi_visualizer.py acem--ilahi--duyek--aldanma_dunya--zekai_dede.mid
```

**Comment by hedonistrh** [comment]
In their [midi-io script](https://github.com/tensorflow/magenta/blob/master/magenta/music/midi_io.py#L86), they convert midis into [_ProtoBuffer_](https://developers.google.com/protocol-buffers/) via [pretty-midi library.](https://github.com/craffel/pretty-midi)

I have not any experience with _ProtoBuffer._ So, its manipulation can be hard. **Also, we can convert all our midis into ProtoBuffer without any problem. Our problem is that when we use these protobuffer as an input for MelodyRNN, some of them raise _polyphony exception_** 🙍 

At that point, I suggest that lets drop MelodyRNN at that step. 

- [ ] Close the issue #32 #31, after add some info them.

- [ ] Focus on text-based generation via RNN. You can check my [mock training notebook](https://github.com/hedonistrh/TurkishMusicGeneration/blob/02dda380c3ddce42ccdb304aaf69123a13dab83e/MockTraining.ipynb) or [MIT's music generation lab.](https://github.com/aamini/introtodeeplearning_labs/blob/master/lab1/Part2_music_generation_solution.ipynb) I can convert mock training notebook into scripted version, so we can easily tweak it and runs into cluster. After that, we should add _attention_ and _lookback_ variation of it. :octocat:

**Comment by hedonistrh** [comment]
We can not use _MelodyRNN_. We are going to reproduce their architecture ourselves. We are planning to use _.txt_ files as our input.

**Comment by magdalenafuentes** [comment]
The encoder-decoder interface for Magenta's melody models is [here](https://github.com/tensorflow/magenta/blob/master/magenta/music/melody_encoder_decoder.py), it is basically: 

> MelodyOneHotEncoding is an encoder_decoder.OneHotEncoding that specifies a one-
> hot encoding for Melody events, i.e. MIDI pitch values plus note-off and no-
> event.
> KeyMelodyEncoderDecoder is an encoder_decoder.EventSequenceEncoderDecoder that
> specifies an encoding of Melody objects into input vectors and output labels for
> use by melody models.

So, they are only considering pitch, beginning, end of notes (even though [pretty-midi encodes pitch-bend information  ](http://craffel.github.io/pretty-midi/#pretty-midi-note)).

I guess it's quite clear that the hack of SymbTR will be ignored by magenta models. I couldn't find the paper related to Melody-RNN though, do u have a reference @hedonistrh ?

I agree if you want to move to custom model, just a thought in case is useful here or afterwards: if melodies in SymbTR are contained within 2-3 octaves, maybe it would be possible to encode information of pitch-bend in the other octaves (i.e. mapping the ,X to higher octaves so they live in standard midi notation).

**Comment by hedonistrh** [comment]
Hey @magdalenafuentes According to [this comment](https://github.com/tensorflow/magenta/issues/1438#issuecomment-468747599), there is no paper about Melody-RNN. 

I think that we should not directly focus on standart midi notation. At that time, Koma53 value makes more sense for me.
